diff.go

```diff
@@ -2,154 +2,119 @@ package userconfig
 
 import (
 	"reflect"
+	"sort"
 )
 
 type DiffType string
 
 const (
-	// Used to indicate that the app-name in the two files was changed
-	InfoAppNameChanged DiffType = "infoAppNameChanged"
+	// DiffInfoServiceNameUpdated
+	DiffInfoServiceNameUpdated DiffType = "service-name-updated"
 
-	// Used to indicate that a node (service or component) was added. `Name` in the DiffInfo describes the node that was added.
-	InfoNodeAdded DiffType = "node-added"
+	// DiffInfoComponentAdded
+	DiffInfoComponentAdded DiffType = "component-added"
 
-	// Used to indicate that a node (service or component) was removed. `Name` in the DiffInfo describes the node that was removed.
-	InfoNodeRemoved DiffType = "node-removed"
+	// DiffInfoComponentRemoved
+	DiffInfoComponentRemoved DiffType = "component-removed"
 
-	// Used to indicate that the scaling config of a component changed.
-	InfoComponentScalingUpdated DiffType = "component-scaling-changed"
-
-	// Used to indicate that the InstanceConfig in the ComponentConfig identified by `Name` changed.
-	InfoInstanceConfigUpdated DiffType = "instance-update"
-
-	// Used to indicate that the component itself changed (but not the underlying InstanceConfig). `Name` in the DiffInfo describes the node that was changed.
-	InfoComponentUpdated DiffType = "component-update"
+	// DiffInfoComponentUpdated
+	DiffInfoComponentUpdated DiffType = "component-updated"
 )
 
 type DiffInfo struct {
-	// What type changed: app, service, component
 	Type DiffType
 
-	// Path to the node that changed.
-	Name []string
+	Old string
+	New string
 }
 
-// Diff compares the two AppConfigs and returns a list of changes between the two.
-func Diff(newConfig, oldConfig AppDefinition) []DiffInfo {
-	if newConfig.AppName != oldConfig.AppName {
-		return []DiffInfo{
-			DiffInfo{Type: InfoAppNameChanged, Name: []string{oldConfig.AppName}},
-		}
-	}
-
-	changes := []DiffInfo{}
-	c := make(chan DiffInfo)
-	go func() {
-		diffHierarchy([]string{newConfig.AppName}, appNode(newConfig), appNode(oldConfig), c)
-		close(c)
-	}()
-
-	for change := range c {
-		changes = append(changes, change)
-	}
-	return changes
-}
+func Diff(oldDef, newDef V2AppDefinition) []DiffInfo {
+	diffInfos := []DiffInfo{}
 
-type node interface {
-	Name() string
-	Children() []node
+	diffInfos = append(diffInfos, diffServiceNameUpdated(newDef.AppName, oldDef.AppName)...)
+	diffInfos = append(diffInfos, diffComponentAdded(newDef.Components, oldDef.Components)...)
+	diffInfos = append(diffInfos, diffComponentUpdated(newDef.Components, oldDef.Components)...)
+	diffInfos = append(diffInfos, diffComponentRemoved(newDef.Components, oldDef.Components)...)
 
-	// Diff is called by diffHierarchy when two nodes have the same name and should be checked
-	// for internal differences. Diffs should be written to changes.
-	Diff(path []string, other node, changes chan<- DiffInfo)
+	return diffInfos
 }
 
-type appNode AppDefinition
+func diffServiceNameUpdated(a, b AppName) []DiffInfo {
```

Comment:
I don't like `a` and `b` as argument names. Please change to `old` and `new` or something like that.